### PR TITLE
fix: upgrade typescript + prettier + eslint with a single shared linting config

### PR DESCRIPTION
**Pull request by stefl**
## Description

- Upgrades Typescript to latest
- Upgrades Eslint to latest
- Moves all config to a shared eslint-config package
- Uses the new Eslint config format

We primarily use Typescript, some ESM, and an occasional CommonJS file. So this makes TS the preferred format, ESM the secondary and CommonJS a tertiary format.

To write JS code, name your file with the .mjs extension.
To write code with CommonJS you will need to use the .cjs extension.

Linting rules will now apply to each type of file, and we can specify rules that apply generally, or specifically for certain types of file.

## How to review

- Pull the branch down locally
- Restart your editor
- Make changes to a file / break something
- Make sure that linting is performant and you get feedback as expected
- Fix the issue
- Make sure you get feedback quickly that the error is resolved

### CODE REVIEW

File: apps/nextjs/package.json [hunk 1]
```diff
@@ -150,17 +152,14 @@
...
     "graphql": "^16.9.0",
     "jest": "^29.7.0",
     "msw": "^2.6.5",
     "msw-storybook-addon": "^2.0.4",
     "postcss": "^8.4.32",
     "tailwindcss": "^3.3.7",
-    "ts-jest": "^29.1.4",
-    "typescript": "5.3.3",
+    "ts-jest": "^29.2.5",
+    "typescript": "5.7.2",
```
Comment (stefl) [inline apps/nextjs/package.json:162]: Bumps Typescript to latest

File: apps/nextjs/src/components/AppComponents/Chat/chat-message/index.tsx [hunk 2]
```diff
@@ -1,6 +1,6 @@
 // Inspired by Chatbot-UI and modified to fit the needs of this project
 // @see https://github.com/mckaywrigley/chatbot-ui/blob/main/components/Chat/ChatMessage.tsx
-import type { ReactNode } from "react";
+import type { ReactNode, JSX } from "react";
```
Comment (stefl) [inline apps/nextjs/src/components/AppComponents/Chat/chat-message/index.tsx:3]: Global JSX is now deprecated

File: apps/nextjs/tests-e2e/tests/aila-chat/helpers.ts [hunk 3]
```diff
@@ -21,7 +21,7 @@ export async function waitForStreamingStatusChange(
   await page.waitForFunction(
     ([currentStatus, expectedStatus]) => {
       const statusElement = document.querySelector(
-        // eslint-disable-next-line @typescript-eslint/quotes, quotes
+        // eslint-disable-next-line quotes
```
Comment (stefl) [inline apps/nextjs/tests-e2e/tests/aila-chat/helpers.ts:24]: `@typescript-eslint/quotes` is deprecated

File: apps/nextjs/tsconfig.json [hunk 4]
```diff
@@ -9,11 +9,13 @@
     "incremental": true,
     "esModuleInterop": true,
     "module": "esnext",
-    "moduleResolution": "node",
+    "moduleResolution": "bundler",
```
Comment (stefl) [inline apps/nextjs/tsconfig.json:12]: Move back to Bundler throughout

File: packages/api/package.json [hunk 5]
```diff
@@ -4,12 +4,20 @@
...
   "types": "./index.ts",
+
   "scripts": {
     "check": "tsc --noEmit",
     "clean": "rm -rf .turbo node_modules",
-    "lint": "eslint . --ext .ts,.tsx",
+    "lint": "eslint .",
     "type-check": "tsc --noEmit"
   },
+  "eslintConfig": {
```
Comment (stefl) [inline packages/api/package.json:14]: Remove the per-project config (for now?) and replace with this config in the package.json for each package

File: packages/api/package.json [hunk 6]
```diff
@@ -26,12 +34,13 @@
...
-    "eslint": "^8.56.0",
-    "typescript": "5.3.3"
+    "typescript": "5.7.2"
   },
   "engines": {
     "node": ">=20.9.0",
     "pnpm": ">=8"
-  }
+  },
+  "type": "module"
```
Comment (stefl) [inline packages/api/package.json:45]: Set all packages as type module

File: apps/nextjs/src/app/api/chat/fixtures/FixtureRecordOpenAiClient.ts [hunk 7]
```diff
@@ -3,7 +3,7 @@ import { createOpenAIClient } from "@oakai/core/src/llm/openai";
 import { aiLogger } from "@oakai/logger";
 import fs from "fs/promises";
 import type OpenAI from "openai";
-import type { ChatCompletionCreateParamsNonStreaming } from "openai/resources";
+import type { ChatCompletionCreateParamsNonStreaming } from "openai/resources/index.mjs";
```
Comment (stefl) [inline apps/nextjs/src/app/api/chat/fixtures/FixtureRecordOpenAiClient.ts:6]: I'm not sure why we have to do this now. This has changed a few times

File: apps/nextjs/src/lib/hooks/use-sidebar.tsx [hunk 8]
```diff
@@ -11,13 +11,15 @@ import {
...
 
-export interface SidebarContext {
+export interface SidebarContextValue {
   isSidebarOpen: boolean;
   toggleSidebar: () => void;
   isLoading: boolean;
 }
 
-const SidebarContext = createContext<SidebarContext | undefined>(undefined);
+const SidebarContext = createContext<SidebarContextValue | undefined>(
```
Comment (stefl) [inline apps/nextjs/src/lib/hooks/use-sidebar.tsx:20]: The eslint config no longer likes types named the same as code definitions

File: packages/eslint-config/package.json [hunk 9]
```diff
@@ -0,0 +1,42 @@
+{
+  "name": "@oakai/eslint-config",
```
Comment (stefl) [inline packages/eslint-config/package.json:2]: This makes a new package that matches the same naming convention as the prettier config

File: packages/eslint-config/src/index.mjs [hunk 10]
```diff
@@ -0,0 +1,140 @@
+import eslint from '@eslint/js';
```
Comment (stefl) [inline packages/eslint-config/src/index.mjs:1]: This is the single source of truth for eslint config. ATM it has all projects in one file

File: packages/eslint-config/src/plugins.d.ts [hunk 11]
```diff
@@ -0,0 +1,3 @@
+declare module 'eslint-plugin-import';
```
Comment (stefl) [inline packages/eslint-config/src/plugins.d.ts:1]: These don't have types

File: packages/eslint-config/src/rules.mjs [hunk 12]
```diff
@@ -0,0 +1,64 @@
+import tsEslint from '@typescript-eslint/eslint-plugin';
```
Comment (stefl) [inline packages/eslint-config/src/rules.mjs:1]: Separate out the actual rule definitions

File: packages/ingest/src/steps/5-lp-parts-embed-start.ts [hunk 13]
```diff
@@ -78,7 +78,9 @@ export async function lpPartsEmbedStart({
           await prisma.ingestLessonPlanPart.updateMany({
             where: {
               id: {
-                in: lessonPlanPartIds,
+                in: lessonPlanPartIds.filter(
```
Comment (stefl) [inline packages/ingest/src/steps/5-lp-parts-embed-start.ts:81]: Fixes a linting error

File: packages/prettier-config/package.json [hunk 14]
```diff
@@ -1,15 +1,14 @@
...
-  "main": "index.js",
-  "dependencies": {
-    "@trivago/prettier-plugin-sort-imports": "^4.3.0",
-    "prettier-plugin-tailwindcss": "^0.5.9"
-  },
+  "main": "index.mjs",
   "devDependencies": {
-    "prettier": "^3.1.1"
+    "@trivago/prettier-plugin-sort-imports": "^4.3.0",
+    "prettier": "^3.4.1",
```
Comment (stefl) [inline packages/prettier-config/package.json:7]: Bump prettier to latest

File: .github/actions/ref_from_sha/test/e2e.cjs [hunk 15]
```diff
@@ -1,7 +1,7 @@
 const github = require("@actions/github");
 
-const prFromSha = require("../pr_from_sha");
-const branchFromSha = require("../branch_from_sha");
+const prFromSha = require("../pr_from_sha.cjs");
```
Comment (stefl) [inline .github/actions/ref_from_sha/test/e2e.cjs:3]: Potentially we could refactor this all to use ESM but this is the minimal change to get this to work with the new requirement to use .cjs extensions for commonjs code

File: apps/nextjs/package.json [hunk 16]
```diff
@@ -1,26 +1,27 @@
...
   "name": "@oakai/nextjs",
-  "version": "0.1.0",
+  "version": "1.0.0",
   "private": true,
   "scripts": {
     "build": "next build",
     "build:dev": "pnpm with-env next build",
     "check": "tsc --noEmit",
     "clean": "rm -rf .next .turbo node_modules",
-    "dev": "FORCE_COLOR=1 SENTRY_SUPPRESS_TURBOPACK_WARNING=1 pnpm with-env node scripts/increase-listeners.js next dev --port 2525 --turbo | pino-pretty -C",
```
Comment (codeincontext) [inline apps/nextjs/package.json:10]: Nice!

File: apps/nextjs/package.json [hunk 17]
```diff
@@ -1,26 +1,27 @@
 {
   "name": "@oakai/nextjs",
-  "version": "0.1.0",
+  "version": "1.0.0",
```
Comment (codeincontext) [inline apps/nextjs/package.json:3]: Lovely

**Review by codeincontext — APPROVED** [review]
The code looks great to me. I haven't tested it locally yet

**Review by codeincontext — APPROVED** [review]
Tested and it seems to work very well with Zed (my editor)

**Pull request merged**